libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/Trace.hs

```diff
@@ -0,0 +1,283 @@
+    EpochState (..),
+    LedgerState (..),
+    NewEpochState (..),
+  )
+import qualified Cardano.Ledger.Shelley.PParams as Shelley (PParams' (..))
+import Cardano.Slotting.Slot (EpochNo (..), SlotNo (..))
+import Control.Monad.Trans.Class (MonadTrans (lift))
+import Control.Monad.Trans.RWS.Strict (get, gets)
+import Control.Monad.Trans.Reader (ReaderT (..))
+import Control.State.Transition.Extended (STS (..))
+import Control.State.Transition.Trace (Trace (..))
+import Control.State.Transition.Trace.Generator.QuickCheck (HasTrace (..), traceFromInitState)
+import Data.Default.Class (Default (def))
+import Data.Functor.Identity (Identity (runIdentity))
+import qualified Data.Map as Map
+import Data.Maybe.Strict (StrictMaybe (..))
+import Data.Sequence as Seq hiding (length, reverse)
+import Data.Vector (Vector, (!))
+import qualified Data.Vector as Vector
+import GHC.Word (Word64)
+import Test.Cardano.Ledger.Generic.ApplyTx (applyTx)
+import Test.Cardano.Ledger.Generic.GenState
+  ( GenEnv (..),
+    GenRS,
+    GenSize,
+    GenState (..),
+    getBlocksizeMax,
+    getReserves,
+    getSlot,
+    getTreasury,
+    initialLedgerState,
+    modifyModel,
+    runGenRS,
+  )
+import Test.Cardano.Ledger.Generic.MockChain
+import Test.Cardano.Ledger.Generic.ModelState
+  ( Model,
+    mNewEpochStateZero,
+    stashedAVVMAddressesZero,
+  )
+import Test.Cardano.Ledger.Generic.PrettyCore (pcTx)
+import Test.Cardano.Ledger.Generic.Proof hiding (lift)
+import Test.Cardano.Ledger.Generic.TxGen
+  ( genValidatedTx,
+  )
+import Test.Cardano.Ledger.Shelley.Utils (testGlobals)
+import Test.QuickCheck
+import Test.Tasty (TestTree, testGroup)
+import Test.Tasty.HUnit (testCaseSteps)
+
+-- ===========================================
+
+-- | Generate a Core.Tx and an internal Model of the state after the block
+--   has been applied. That model can be used to generate the next Tx
+genRsTxAndModel :: Reflect era => Proof era -> Model era -> Word64 -> GenRS era (Core.Tx era, Model era)
+genRsTxAndModel proof model0 n = do
+  modifyModel (const model0)
+  (_, tx) <- genValidatedTx proof
+  model1 <- gets gsModel
+  let model2 = applyTx proof (fromIntegral n) model1 tx
+  pure (tx, model2)
+
+-- | Generate a Vector of [(Word64,Core.Tx era)] representing a (Vector Block)
+genRsTxSeq ::
+  forall era.
+  Reflect era =>
+  Proof era ->
+  Model era ->
+  Word64 ->
+  Word64 ->
+  [(Word64, Core.Tx era)] ->
+  GenRS era (Vector [(Word64, Core.Tx era)])
+genRsTxSeq _ _ this lastN ans | this > lastN = chop (reverse ans) []
+genRsTxSeq proof model0 this lastN ans = do
+  (tx, model1) <- genRsTxAndModel proof model0 this
+  genRsTxSeq proof model1 (this + 1) lastN ((this, tx) : ans)
+
+-- | Chop a list into random size blocks
+chop :: [(Word64, Core.Tx era)] -> [[(Word64, Core.Tx era)]] -> GenRS era (Vector [(Word64, Core.Tx era)])
+chop [] ans = pure $ Vector.fromList (reverse ans)
+chop xs ans | length xs <= 3 = pure $ Vector.fromList (reverse (xs : ans))
```

Comment:
That line can be quite expensive on large `xs`
```suggestion
chop xs ans | length (take 4 xs) <= 3 = pure $ Vector.fromList (reverse (xs : ans))
```

Reply:
pretty clever